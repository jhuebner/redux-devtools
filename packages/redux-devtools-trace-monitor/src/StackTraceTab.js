import React, { Component } from 'react';

import { getStackFrames } from './react-error-overlay/utils/getStackFrames';
import StackTrace from './react-error-overlay/containers/StackTrace';
import openFile from './openFile';

const rootStyle = { padding: '5px 10px' };

export default class StackTraceTab extends Component {
  static defaultProps = {
    openFile
  };
  constructor(props) {
    super(props);

    this.state = {
      stackFrames: []
    };
  }
  componentDidMount() {
    // console.log("StackTraceTab mounted");
    this.checkForStackTrace();
  }

  componentDidUpdate(prevProps) {
    const { action, actions } = prevProps;

    if (action !== this.props.action || actions !== this.props.actions) {
      this.checkForStackTrace();
    }
  }

  checkForStackTrace() {
    const { action, actions: liftedActionsById } = this.props;

    if (!action) {
      return;
    }

    const liftedActions = Object.values(liftedActionsById);
    const liftedAction = liftedActions.find(
      liftedAction => liftedAction.action === action
    );

    if (liftedAction && typeof liftedAction.stack === 'string') {
      const deserializedError = Object.assign(new Error(), {
        stack: liftedAction.stack
      });

      getStackFrames(deserializedError).then(stackFrames => {
        /* eslint-disable no-console */
        if (process.env.NODE_ENV === 'development')
          console.log('Stack frames: ', stackFrames);
        /* eslint-enable no-console */
        this.setState({ stackFrames, currentError: deserializedError });
      });
    } else {
      this.setState({
        stackFrames: [],
        showDocsLink:
          liftedAction.action &&
          liftedAction.action.type &&
          liftedAction.action.type !== '@@INIT'
      });
    }
  }

  onStackLocationClicked = (fileLocation = {}) => {
    // console.log("Stack location args: ", ...args);

    const { fileName, lineNumber } = fileLocation;

    if (fileName && lineNumber) {
      const matchingStackFrame = this.state.stackFrames.find(stackFrame => {
        const matches =
          (stackFrame._originalFileName === fileName &&
            stackFrame._originalLineNumber === lineNumber) ||
          (stackFrame.fileName === fileName &&
            stackFrame.lineNumber === lineNumber);
        return matches;
      });

      // console.log("Matching stack frame: ", matchingStackFrame);

      if (matchingStackFrame) {
        /*
        const frameIndex = this.state.stackFrames.indexOf(matchingStackFrame);
        const originalStackFrame = parsedFramesNoSourcemaps[frameIndex];
        console.log("Original stack frame: ", originalStackFrame);
        */
        this.props.openFile(fileName, lineNumber, matchingStackFrame);
      }
    }
  };

  openDocs = e => {
    e.stopPropagation();
    window.open(
      'https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/Features/Trace.md'
    );
  };

  render() {
    const { stackFrames, showDocsLink } = this.state;

    if (showDocsLink) {
      return (
        <div style={rootStyle}>
          To enable tracing action calls, you should set `trace` option to
          `true` for Redux DevTools enhancer. Refer to{' '}
          <a href="#" onClick={this.openDocs}>
            this page
          </a>{' '}
          for more details.
        </div>
      );
    }

    return (
      <div style={rootStyle}>
        <StackTrace
          stackFrames={stackFrames}
          errorName="N/A"
          contextSize={3}
          editorHandler={this.onStackLocationClicked}
        />
      </div>
    );
  }
}
