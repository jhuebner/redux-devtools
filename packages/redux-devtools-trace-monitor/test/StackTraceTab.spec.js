import React from 'react';
import { configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import StackTraceTab from '../src/StackTraceTab';

import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });

function genAsyncSnapshot(component, done) {
  setTimeout(() => {
    component.update();
    expect(toJson(component)).toMatchSnapshot();
    done();
  });
}

const actions = {
  0: { type: 'PERFORM_ACTION', action: { type: '@@INIT' } },
  1: { type: 'PERFORM_ACTION', action: { type: 'INCREMENT_COUNTER' } },
  2: {
    type: 'PERFORM_ACTION',
    action: { type: 'INCREMENT_COUNTER' },
    stack:
      'Error\n    at fn1 (app.js:72:24)\n    at fn2 (app.js:84:31)\n     ' +
      'at fn3 (chrome-extension://lmhkpmbekcpmknklioeibfkpmmfibljd/js/page.bundle.js:1269:80)'
  }
};

describe('StackTraceTab component', () => {
  it('should render with no props', done => {
    const component = mount(<StackTraceTab />);
    genAsyncSnapshot(component, done);
  });

  it('should render with props, but without stack', done => {
    const component = mount(
      <StackTraceTab actions={actions} action={actions[0].action} />
    );
    genAsyncSnapshot(component, done);
  });

  it('should render the link to docs', done => {
    const component = mount(
      <StackTraceTab actions={actions} action={actions[1].action} />
    );
    genAsyncSnapshot(component, done);
  });

  it('should render with trace stack', done => {
    const component = mount(
      <StackTraceTab actions={actions} action={actions[2].action} />
    );
    genAsyncSnapshot(component, done);
  });
});
