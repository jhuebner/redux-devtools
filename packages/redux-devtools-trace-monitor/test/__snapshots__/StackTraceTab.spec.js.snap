// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StackTraceTab component should render the link to docs 1`] = `
<StackTraceTab
  action={
    Object {
      "type": "INCREMENT_COUNTER",
    }
  }
  actions={
    Object {
      "0": Object {
        "action": Object {
          "type": "@@INIT",
        },
        "type": "PERFORM_ACTION",
      },
      "1": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "type": "PERFORM_ACTION",
      },
      "2": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "stack": "Error
    at fn1 (app.js:72:24)
    at fn2 (app.js:84:31)
     at fn3 (chrome-extension://lmhkpmbekcpmknklioeibfkpmmfibljd/js/page.bundle.js:1269:80)",
        "type": "PERFORM_ACTION",
      },
    }
  }
  openFile={[Function]}
>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    To enable tracing action calls, you should set \`trace\` option to \`true\` for Redux DevTools enhancer. Refer to
     
    <a
      href="#"
      onClick={[Function]}
    >
      this page
    </a>
     
    for more details.
  </div>
</StackTraceTab>
`;

exports[`StackTraceTab component should render with no props 1`] = `
<StackTraceTab
  openFile={[Function]}
>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <StackTrace
      contextSize={3}
      editorHandler={[Function]}
      errorName="N/A"
      stackFrames={Array []}
    >
      <div
        style={
          Object {
            "flex": "0 1 auto",
            "fontSize": "1em",
            "minHeight": "0px",
            "overflow": "auto",
          }
        }
      />
    </StackTrace>
  </div>
</StackTraceTab>
`;

exports[`StackTraceTab component should render with props, but without stack 1`] = `
<StackTraceTab
  action={
    Object {
      "type": "@@INIT",
    }
  }
  actions={
    Object {
      "0": Object {
        "action": Object {
          "type": "@@INIT",
        },
        "type": "PERFORM_ACTION",
      },
      "1": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "type": "PERFORM_ACTION",
      },
      "2": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "stack": "Error
    at fn1 (app.js:72:24)
    at fn2 (app.js:84:31)
     at fn3 (chrome-extension://lmhkpmbekcpmknklioeibfkpmmfibljd/js/page.bundle.js:1269:80)",
        "type": "PERFORM_ACTION",
      },
    }
  }
  openFile={[Function]}
>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <StackTrace
      contextSize={3}
      editorHandler={[Function]}
      errorName="N/A"
      stackFrames={Array []}
    >
      <div
        style={
          Object {
            "flex": "0 1 auto",
            "fontSize": "1em",
            "minHeight": "0px",
            "overflow": "auto",
          }
        }
      />
    </StackTrace>
  </div>
</StackTraceTab>
`;

exports[`StackTraceTab component should render with trace stack 1`] = `
<StackTraceTab
  action={
    Object {
      "type": "INCREMENT_COUNTER",
    }
  }
  actions={
    Object {
      "0": Object {
        "action": Object {
          "type": "@@INIT",
        },
        "type": "PERFORM_ACTION",
      },
      "1": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "type": "PERFORM_ACTION",
      },
      "2": Object {
        "action": Object {
          "type": "INCREMENT_COUNTER",
        },
        "stack": "Error
    at fn1 (app.js:72:24)
    at fn2 (app.js:84:31)
     at fn3 (chrome-extension://lmhkpmbekcpmknklioeibfkpmmfibljd/js/page.bundle.js:1269:80)",
        "type": "PERFORM_ACTION",
      },
    }
  }
  openFile={[Function]}
>
  <div
    style={
      Object {
        "padding": "5px 10px",
      }
    }
  >
    <StackTrace
      contextSize={3}
      editorHandler={[Function]}
      errorName="N/A"
      stackFrames={
        Array [
          StackFrame {
            "_originalColumnNumber": null,
            "_originalFileName": null,
            "_originalFunctionName": null,
            "_originalLineNumber": null,
            "_originalScriptCode": null,
            "_scriptCode": null,
            "columnNumber": 24,
            "fileName": "app.js",
            "functionName": "fn1",
            "lineNumber": 72,
          },
          StackFrame {
            "_originalColumnNumber": null,
            "_originalFileName": null,
            "_originalFunctionName": null,
            "_originalLineNumber": null,
            "_originalScriptCode": null,
            "_scriptCode": null,
            "columnNumber": 31,
            "fileName": "app.js",
            "functionName": "fn2",
            "lineNumber": 84,
          },
        ]
      }
    >
      <div
        style={
          Object {
            "flex": "0 1 auto",
            "fontSize": "1em",
            "minHeight": "0px",
            "overflow": "auto",
          }
        }
      >
        <Collapsible
          collapsedByDefault={false}
          key="bundle-1"
        >
          <div>
            <button
              onClick={[Function]}
              style={
                Object {
                  "backgroundColor": "#3C444F",
                  "border": "none",
                  "color": "#FFFFFF",
                  "cursor": "pointer",
                  "display": "block",
                  "fontSize": "1em",
                  "lineHeight": "1.5",
                  "marginBottom": "0.6em",
                  "padding": "0px 5px",
                  "textAlign": "left",
                  "width": "100%",
                }
              }
            >
              ▼ 2 stack frames were expanded.
            </button>
            <div
              style={
                Object {
                  "display": "block",
                }
              }
            >
              <StackFrame
                contextSize={3}
                critical={true}
                editorHandler={[Function]}
                frame={
                  StackFrame {
                    "_originalColumnNumber": null,
                    "_originalFileName": null,
                    "_originalFunctionName": null,
                    "_originalLineNumber": null,
                    "_originalScriptCode": null,
                    "_scriptCode": null,
                    "columnNumber": 24,
                    "fileName": "app.js",
                    "functionName": "fn1",
                    "lineNumber": 72,
                  }
                }
                key="frame-0"
                showCode={false}
              >
                <div>
                  <div>
                    fn1
                  </div>
                  <div
                    style={
                      Object {
                        "fontSize": "0.9em",
                        "marginBottom": "0.9em",
                      }
                    }
                  >
                    <span
                      onClick={null}
                      onKeyDown={null}
                      style={null}
                      tabIndex={null}
                    >
                      app.js:72:24
                    </span>
                  </div>
                </div>
              </StackFrame>
              <StackFrame
                contextSize={3}
                critical={false}
                editorHandler={[Function]}
                frame={
                  StackFrame {
                    "_originalColumnNumber": null,
                    "_originalFileName": null,
                    "_originalFunctionName": null,
                    "_originalLineNumber": null,
                    "_originalScriptCode": null,
                    "_scriptCode": null,
                    "columnNumber": 31,
                    "fileName": "app.js",
                    "functionName": "fn2",
                    "lineNumber": 84,
                  }
                }
                key="frame-1"
                showCode={false}
              >
                <div>
                  <div>
                    fn2
                  </div>
                  <div
                    style={
                      Object {
                        "fontSize": "0.9em",
                        "marginBottom": "0.9em",
                      }
                    }
                  >
                    <span
                      onClick={null}
                      onKeyDown={null}
                      style={null}
                      tabIndex={null}
                    >
                      app.js:84:31
                    </span>
                  </div>
                </div>
              </StackFrame>
              <button
                onClick={[Function]}
                style={
                  Object {
                    "backgroundColor": "#3C444F",
                    "border": "none",
                    "color": "#FFFFFF",
                    "cursor": "pointer",
                    "display": "block",
                    "fontSize": "1em",
                    "lineHeight": "1.5",
                    "marginBottom": "0.6em",
                    "padding": "0px 5px",
                    "textAlign": "left",
                    "width": "100%",
                  }
                }
              >
                ▲ 2 stack frames were expanded.
              </button>
            </div>
          </div>
        </Collapsible>
      </div>
    </StackTrace>
  </div>
</StackTraceTab>
`;
